File: common/src/main/java/org/kryptonmc/nbt/StringTag.java
/*
 * This file is part of Krypton NBT, licensed under the MIT license.
 *
 * Copyright (C) 2021 KryptonMC and contributors
 *
 * This project is licensed under the terms of the MIT license.
 * For more details, please reference the LICENSE file in the top-level directory.
 */
package org.kryptonmc.nbt;

import org.jetbrains.annotations.NotNull;

/**
 * A tag that holds a string value.
 */
public sealed interface StringTag extends Tag permits StringTagImpl {

    /**
     * The string tag representing the empty string.
     */
    @NotNull StringTag EMPTY = new StringTagImpl("");

    /**
     * The ID of this type of tag.
     *
     * <p>Used for {@link CollectionTag#elementType()} and in the serialized
     * binary form.</p>
     */
    int ID = 8;
    /**
     * The tag type for this tag.
     */
    @NotNull TagType<@NotNull StringTag> TYPE = StringTagImpl.createType();

    /**
     * Gets a string tag that represents the given value.
     *
     * @param value the backing value
     * @return a string tag representing the value
     */
    static @NotNull StringTag of(final @NotNull String value) {
        return value.isEmpty() ? EMPTY : new StringTagImpl(value);
    }

    /**
     * Gets the underlying value of this string tag.
     *
     * @return the value
     */
    @NotNull String value();

    @Override
    @NotNull StringTag copy();
}

File: common/src/main/kotlin/org/kryptonmc/nbt/StringTag.kt
/*
 * This file is part of Krypton NBT, licensed under the MIT license.
 *
 * Copyright (C) 2021 KryptonMC and contributors
 *
 * This project is licensed under the terms of the MIT license.
 * For more details, please reference the LICENSE file in the top-level directory.
 */
package org.kryptonmc.nbt

import org.kryptonmc.nbt.io.TagReader
import org.kryptonmc.nbt.io.TagWriter
import java.io.DataOutput
import kotlin.jvm.JvmField
import kotlin.jvm.JvmStatic

public class StringTag private constructor(public val value: String) : Tag {

    override val id: Int
        get() = ID
    override val type: TagType
        get() = TYPE

    override fun write(output: DataOutput) {
        WRITER.write(output, this)
    }

    override fun <T> examine(examiner: TagExaminer<T>) {
        examiner.examineString(this)
    }

    override fun copy(): StringTag = this

    override fun asString(): String = value

    override fun equals(other: Any?): Boolean = this === other || (other is StringTag && value == other.value)

    override fun hashCode(): Int = value.hashCode()

    override fun toString(): String = value

    public companion object {

        public const val ID: Int = 8
        @JvmField
        public val TYPE: TagType = TagType("TAG_String", true)
        @JvmField
        public val READER: TagReader<StringTag> = TagReader { input, _ -> of(input.readUTF()) }
        @JvmField
        public val WRITER: TagWriter<StringTag> = TagWriter { output, value -> output.writeUTF(value.value) }
        private val EMPTY = StringTag("")

        /**
         * Creates a new string tag holding the given [value].
         *
         * @param value the value
         * @return a new string tag
         */
        @JvmStatic
        public fun of(value: String): StringTag {
            if (value.isEmpty()) return empty()
            return StringTag(value)
        }

        /**
         * Gets the string tag that holds the empty string.
         *
         * @return the empty string tag
         */
        @JvmStatic
        public fun empty(): StringTag = EMPTY
    }
}
